### PR TITLE
[pkg/util/log] Add zap wrapper over agent logger

### PR DESCRIPTION
**Pull request by mx-psi**
### What does this PR do?

Add [`zap`](https://pkg.go.dev/go.uber.org/zap) wrapper over the Agent's default logger.

### Motivation

Part of DataDog/architecture#892, to be able to log messages from OpenTelemetry Collector components.

### Additional Notes

We adapt some code from the zap library which is sadly not exported.

### Describe how to test your changes

Test that existing log code works properly (specially file, line and function should be correct)

### Checklist

- [x] A [release note](https://github.com/DataDog/datadog-agent/blob/main/docs/dev/contributing.md#reno) has been added or the `changelog/no-changelog` label has been applied.
- [x] The `need-change/operator` and `need-change/helm` labels has been applied if applicable.
- [x] The appropriate `team/..` label has been applied, if known.
- [x] If known, an appropriate milestone has been selected; otherwise the `Triage` milestone is set.
- [x] The [config template](https://github.com/DataDog/datadog-agent/blob/main/pkg/config/config_template.yaml) has been updated if applicable.

Note: Adding GitHub labels is only possible for contributors with write access.

### CODE REVIEW

File: pkg/util/log/zapcore_test.go [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+		desc    string
+		f       func(*zap.Logger)
+		level   string
+		pattern string
+	}{
+		{
+			desc:    "Debug (no fields, debug level)",
+			f:       func(l *zap.Logger) { l.Debug("Simple message") },
+			level:   "debug",
+			pattern: "\\[DEBUG\\] Write: \\(log/zapcore_test.go:\\d+\\) \\| Simple message",
```
Comment (mx-psi) [inline pkg/util/log/zapcore_test.go:33]: I am not super happy with this way of writing the tests; it works but it's fragile: e.g. if you forget to escape a `|` the tests can easily pass on incorrect situations. I could just hardcode the line numbers which would be safer but more cumbersome to change.

File: pkg/util/log/zapencoder.go [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+	return nil
+}
+
+func (e *encoder) OpenNamespace(ns string) {
+	if e.prefix == "" {
+		e.prefix = ns + "/"
+	} else {
+		e.prefix = e.prefix + ns + "/"
+	}
+}
```
Comment (KSerrania) [inline pkg/util/log/zapencoder.go:79]: Would the following be equivalent to the current implementation:
```suggestion
func (e *encoder) OpenNamespace(ns string) {
	e.prefix = e.prefix + ns + "/"
}
```

given that `e.prefix + ns == ns` if `e.prefix == ""`?

File: pkg/util/log/zapencoder_test.go [hunk 3]
```diff
@@ -0,0 +1,154 @@
...
+		},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.desc, func(t *testing.T) {
+			enc := &encoder{}
+			tt.f(enc)
+			assert.Equal(t, tt.expected, enc.ctx, "Unexpected encoder output.")
+		})
+	}
```
Comment (KSerrania) [inline pkg/util/log/zapencoder_test.go:152]: Same here (renaming `tt` and `f`)

File: pkg/util/log/zapcore_test.go [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+			require.Nil(t, err)
+			SetupLogger(l, tt.level)
+			require.NotNil(t, logger)
+
+			tt.f(logger)
+			w.Flush()
+			pattern := fmt.Sprintf("^%s$", tt.pattern)
+			assert.Regexp(t, pattern, strings.TrimSuffix(b.String(), "\n"))
+		})
+	}
```
[the remark this answers is not in the document]
Comment (mx-psi) [reply pkg/util/log/zapcore_test.go]: Done!

File: pkg/util/log/zapcore_test.go [hunk 5]
```diff
@@ -0,0 +1,131 @@
...
+	logger.Info("Simple message")
+	logger.Info("with fields", zap.Int("val", 1))
+
+	loggerWithCtx := logger.With(zap.String("context", "this"))
+	loggerWithCtx.Warn("A warning", zap.Bool("val", false))
```
[the remark this answers is not in the document]
Comment (mx-psi) [reply pkg/util/log/zapcore_test.go]: It's just that I forgot to remove them 😅, I deleted them now

File: pkg/util/log/zapencoder_test.go (hunk 3 above)
Comment (mx-psi) [reply pkg/util/log/zapencoder_test.go]: Done

**Review by KSerrania — APPROVED** [review]
LGTM

File: pkg/util/log/zapcore.go [hunk 6]
```diff
@@ -0,0 +1,94 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2021-present Datadog, Inc.
+
+package log
```
Comment (olivielpeau) [inline pkg/util/log/zapcore.go:6]: I think `zapcore.go` and `zapencoder.go` would better fit in a new, separate package (maybe `pkg/util/log/zap`?), that would make for a cleaner interface with `pkg/util/log` with exported functions, and avoid adding zap as a dependency of `pkg/util/log`. WDYT?

File: pkg/util/log/log.go [hunk 7]
```diff
@@ -566,9 +570,13 @@ func Tracef(format string, params ...interface{}) {
 	logFormat(seelog.TraceLvl, func() { Tracef(format, params...) }, logger.tracef, format, params...)
 }
 
+func tracecWithDepth(message string, depth int, context ...interface{}) {
```
Comment (olivielpeau) [inline pkg/util/log/log.go:573]: Although I prefer this name, for consistency with the existing functions (example `TraceStackDepth` I think we should call this `tracecStackDepth`), WDYT?

(ditto for all the other similar functions for the other log levels)

File: pkg/util/log/zapcore.go (hunk 6 above)
Comment (mx-psi) [reply pkg/util/log/zapcore.go:6]: That makes sense to me, would we export the `tracecWithDepth`-like functions, then? (so, in your other comment, we would rename e.g. `tracecWithDepth` to `TracecStackDepth`?)

Comment (olivielpeau) [reply pkg/util/log/zapcore.go:6]: Yes, exactly

**Review by olivielpeau — APPROVED** [review]

**Pull request merged**